crates/sargon/src/signing/collector/signatures_collector_preprocessor.rs

```diff
@@ -0,0 +1,149 @@
+            IndexMap::<IntentHash, PetitionForTransaction>::new();
+
+        // We care for only the factor sources which are HD based
+        let mut all_factor_sources_in_profile =
+            HashMap::<FactorSourceIDFromHash, FactorSource>::new();
+        profile_factor_sources.into_iter().for_each(|f| {
+            if let Some(id) = f.factor_source_id().as_hash() {
+                all_factor_sources_in_profile.insert(*id, f);
+            }
+        });
+
+        let mut factor_to_payloads =
+            HashMap::<FactorSourceIDFromHash, IndexSet<IntentHash>>::new();
+
+        let mut used_factor_sources = HashSet::<FactorSource>::new();
+
+        let mut register_factor_in_tx =
+            |id: &FactorSourceIDFromHash, txid: &IntentHash| {
+                if let Some(ref mut txids) = factor_to_payloads.get_mut(id) {
+                    txids.insert(txid.clone());
+                } else {
+                    factor_to_payloads
+                        .insert(*id, IndexSet::just(txid.clone()));
+                }
+
+                assert!(!factor_to_payloads.is_empty());
+
+                let factor_source = all_factor_sources_in_profile
+                    .get(id)
+                    .expect("Should have all factor sources");
+                used_factor_sources.insert(factor_source.clone());
+
+                assert!(!used_factor_sources.is_empty());
+            };
+
+        for transaction in transactions.into_iter() {
+            let mut petitions_for_entities =
+                HashMap::<AddressOfAccountOrPersona, PetitionForEntity>::new();
+
+            for entity in transaction.entities_requiring_auth() {
+                let address = entity.address();
+                match entity.entity_security_state() {
+                    EntitySecurityState::Unsecured { value } => {
+                        let factor_instance = value.transaction_signing; // TODO should change according to payload
+
+                        let factor_source_id = factor_instance.factor_source_id;
+                        register_factor_in_tx(
+                            &factor_source_id,
+                            &transaction.intent_hash,
+                        );
+
+                        let petition = PetitionForEntity::new_unsecurified(
+                            transaction.intent_hash.clone(),
+                            address.clone(),
+                            factor_instance,
+                        );
+                        petitions_for_entities
+                            .insert(address.clone(), petition);
+                    }
+                    EntitySecurityState::Securified { value } => {
+                        let general_role =
+                            GeneralRoleWithHierarchicalDeterministicFactorInstances::try_from(
+                                (value.security_structure.matrix_of_factors, RoleKind::Primary) // Should change according to kind
```

Comment:
we should not hard code the role here, it should be a parameter passed to this `SignaturesCollectorPreprocessor`, which in turn should be passed as a parameter to the constructor of the SignaturesCollector. This **will NOT** cascade (too much) for the test code since the tests code uses specific constructors on SignatureCollector, and those special test ctors they can hard code `RoleKind::Primary`.

Reply:
superseded by https://github.com/radixdlt/sargon/pull/235/files#r1798956661